### PR TITLE
DTW and Unfold for whisper models

### PR DESCRIPTION
**Pull request by zhanghuanrong**

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/tensor/dynamic_time_warping_impl.cu [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        }
+    }
+    *result_len_device = max_index_len - static_cast<size_t>(pos);
+  }
+}
+
+size_t GetDynamicTimeWarpingBufferSize(size_t batch, size_t rows, size_t cols, size_t& max_index_len) {
+  max_index_len = rows + cols + 1;
+  size_t cost_buffer_size = ((rows + 1) * (cols + 1));
+  return batch * max_index_len * 2 * sizeof(int32_t) + // two index arrays
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/tensor/dynamic_time_warping_impl.cu:95]: Might need add alignment to 16 bytes for cost/trace buffers for performance.

File: onnxruntime/contrib_ops/cuda/tensor/unfold_impl.cu [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    int64_t tailing_dims_size,
+    int64_t unfold_size,
+    int64_t step_size
+) {
+  int64_t TPB = device_prop.maxThreadsPerBlock;
+  int64_t unfold_dim_size_dst = (unfold_dim_size - unfold_size) / step_size + 1;
+  int64_t N = leading_dims_size * unfold_dim_size_dst * tailing_dims_size * unfold_size;
+  int64_t num_blocks = (N + TPB - 1) / TPB;
+
+  //   int64_t stride_append_dim_dst = 1;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/tensor/unfold_impl.cu:60]: Remove commented code.

**Comment by tianleiwu** [comment]
Please finish the following:
(1) Add shape inference code; 
(2) Add unit test for DynamicTimeWarping.
(3) Build and update contrib operator documents.
(4) Fix build pipelines

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 3]
```diff
@@ -1062,6 +1062,34 @@ ONNX_MS_OPERATOR_SET_SCHEMA(GridSample, 1,
...
+
+ONNX_MS_OPERATOR_SET_SCHEMA(
+    DynamicTimeWarping, 1,
+    OpSchema()
+        .SetDoc("return dynamic time wrapping of shape [2, x] of int32_t")
+        .Input(0, "input", "input tensor, it must be 2D tensor of shape M x N", "F")
+        .Output(0, "output", "Output tensor.", "I")
+        .TypeConstraint("F", {"tensor(float)"}, "Constrain to float tensors.")
+        .TypeConstraint("I", {"tensor(int32)"}, "Constrain to integer types"));
+
```
Comment (tianleiwu) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:1092]: Add shape inference

**Comment by zhanghuanrong** [comment]
in another PR.

**Pull request closed without merging**